mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala

```diff
@@ -289,17 +301,28 @@ class Word2Vec extends Serializable with Logging {
     val expTable = sc.broadcast(createExpTable())
     val bcVocab = sc.broadcast(vocab)
     val bcVocabHash = sc.broadcast(vocabHash)
-
-    val sentences: RDD[Array[Int]] = words.mapPartitions { iter =>
+    // each partition is a collection of sentences, will be translated into arrays of Index integer
+    val sentences: RDD[Array[Int]] = dataset.mapPartitions { sentenceIter =>
       new Iterator[Array[Int]] {
-        def hasNext: Boolean = iter.hasNext
+        var wordIter: Iterator[String] = null
+
+        def hasNext: Boolean = sentenceIter.hasNext || (wordIter != null && wordIter.hasNext)
 
         def next(): Array[Int] = {
           val sentence = ArrayBuilder.make[Int]
           var sentenceLength = 0
-          while (iter.hasNext && sentenceLength < MAX_SENTENCE_LENGTH) {
-            val word = bcVocabHash.value.get(iter.next())
-            word match {
+          // do translation of each word into its index in the vocabulary,
```

Comment:
I understand that this part of the change intends to respect the implied sentence boundaries in the input. I think it can be simpler? One input sentence maps to 1 or more arrays, and the result should be flattened. Something like?

```
// Each input sentence will produce 1 or more Array[Int], so flatMapPartitions
dataset.flatMapPartitions { sentenceIter => 
  // Each sentence will map to 1 or more Array[Int], so map
  sentenceIter.map { sentence =>
    // Sentence of words, some of which map to a hash, so flatMap
    val hashes = sentence.flatMap(bcVocabHash.value.get)
    // break into sequence of at most maxSentenceLength
    hashes.grouped(maxSentenceLength).map(_.toArray)
  }
}
```

I haven't tested it but does that seem like the intent?